### DISCUSSION TITLE
Fresh install - Incorrect groups auto-increment configuration!

### DISCUSSION THREAD

**Discussion by mightymicke** [requarks/wiki#3358] [Help / Questions]
**Describe the bug**
When run the setup I get this error msg,
"Incorrect groups auto-increment configuration! Should start at 0 and increment by 1. Contact your database administrator".

**To Reproduce**
1. Fresh install the lastest version
2. Start the server.
3. Access the setup guide via webgui on port 3000
4. Add the information and press "install"

**Screenshots**
![image](https://user-images.githubusercontent.com/38280751/74647599-3bd79480-517c-11ea-8aae-f5df9b99dc1a.png)

**Host Info (please complete the following information):**
 - OS: CentOS 8
 - Wiki.js version: 2.1.113
 - Database engine: 10.3.17-MariaDB MariaDB Server

**Comment by NGPixel** [comment]
Does your MariaDB server have a custom configuration for auto-increment?

**Comment by mightymicke** [comment]
I don't know how to change it, or if I can have an custom conf for AUTO_INCREMENT. 

This info is from "MariaDB Server Documentation",
The AUTO_INCREMENT attribute can be used to generate a unique identity for new rows. When you insert a new record to the table, and the auto_increment field is NULL or DEFAULT, the value will automatically be incremented. This also applies to 0, unless the NO_AUTO_VALUE_ON_ZERO SQL_MODE is enabled.

AUTO_INCREMENT columns start from 1 by default. The automatically generated value can never be lower than 0.

https://mariadb.com/kb/en/auto_increment/

**Comment by jhd85** [comment]
I also seem to have this same issue. I have a mariadb galera cluster version 10.4.12. Did not do anything to the auto increment setting. Are there other setting than NO_AUTO_VALUE_ON_ZERO? I do not have NO_AUTO_VALUE_ON_ZERO enabled btw.

**Comment by fenice2** [comment]
I'm also seeing this problem, I'm using Percona MySQL 8 galera cluster 8.0.19-10.1 for the back-end - that's a CentOS 8 server.

**Comment by NiceGuyIT** [comment]
The problem is how Galera handles AUTO_INCREMENT in a cluster. It's best described in [Auto increments in Galera]. TL/DR: if two transactions are started on separate nodes, the first transaction will succeed but the second one will fail because they both received the same AUTO_INCREMENT ID.

The [AUTO_INCREMENT docs] have the following to say:

> To make master-master or Galera safe to use AUTO_INCREMENT one should use the system variables [auto_increment_increment] and [auto_increment_offset] to generate unique values for each server.

This is confirmed by running this on one of the nodes. I'm running a 3 node cluster.

```text
$ echo 'SHOW VARIABLES like "auto_increment%";' | mysql
Variable_name   Value
auto_increment_increment        3
auto_increment_offset   1
```

My first attempt at overcoming this is changing `setup.js` to check for the new IDs.

```js
      if (adminGroup.id !== 3 || guestGroup.id !== 6) {
        throw new Error('Incorrect groups auto-increment configuration! Should start at 0 and increment by 1. Contact your database administrator.')
      }
...
      if (adminUser.id !== 3 || guestUser.id !== 6) {
        throw new Error('Incorrect users auto-increment configuration! Should start at 0 and increment by 1. Contact your database administrator.')
      }
```

That did not work because the application is checking for hard-coded IDs with `findById()`.

```js
  static async getGuestUser () {
    const user = await WIKI.models.users.query().findById(2).withGraphJoined('groups').modifyGraph('groups', builder => {
      builder.select('groups.id', 'permissions')
    })
    if (!user) {
      WIKI.logger.error('CRITICAL ERROR: Guest user is missing!')
      process.exit(1)
    }
    user.permissions = user.getGlobalPermissions()
    return user
  }

  static async getRootUser () {
    let user = await WIKI.models.users.query().findById(1)
    if (!user) {
      WIKI.logger.error('CRITICAL ERROR: Root Administrator user is missing!')
      process.exit(1)
    }
    user.permissions = ['manage:system']
    return user
  }
```

The next attempt was to set the INSERT_ID before inserting the user and groups.

```js
      // Create default groups

      WIKI.logger.info('Creating default groups...')
      await WIKI.models.knex.raw('SET INSERT_ID = 1')
      const adminGroup = await WIKI.models.groups.query().insert({
        name: 'Administrators',
        permissions: JSON.stringify(['manage:system']),
        pageRules: JSON.stringify([]),
        isSystem: true
      })
      await WIKI.models.knex.raw('SET INSERT_ID = 2')
      const guestGroup = await WIKI.models.groups.query().insert({
        name: 'Guests',
        permissions: JSON.stringify(['read:pages', 'read:assets', 'read:comments']),
        pageRules: JSON.stringify([
          { id: 'guest', roles: ['read:pages', 'read:assets', 'read:comments'], match: 'START', deny: false, path: '', locales: [] }
        ]),
        isSystem: true
      })
...

      // Create root administrator
      WIKI.logger.info('Creating root administrator...')
      await WIKI.models.knex.raw('SET INSERT_ID = 1')
      const adminUser = await WIKI.models.users.query().insert({
...

      // Create Guest account
      await WIKI.models.knex.raw('SET INSERT_ID = 2')
      WIKI.logger.info('Creating guest account...')
      const guestUser = await WIKI.models.users.query().insert({
```

This allowed Wiki.js to finish the setup. I was able to login and create the home page. Only time will tell if this solution has unforeseen consequences.

Note: I can add a diff if needed. I figured you can tell the context from the surrounding lines.

[Auto increments in Galera]: https://mariadb.org/auto-increments-in-galera/
[AUTO_INCREMENT docs]: https://mariadb.com/kb/en/auto_increment/#replication
[auto_increment_increment]: https://mariadb.com/kb/en/replication-and-binary-log-system-variables/#auto_increment_increment
[auto_increment_offset]: https://mariadb.com/kb/en/replication-and-binary-log-system-variables/#auto_increment_offset
[INSERT_ID]: https://mariadb.com/kb/en/auto_increment/#setting-or-changing-the-auto_increment-value

**Comment by NGPixel** [comment]
@NiceGuyIT  While the root admin can technically be any ID, the guest user **MUST have ID 2**. Any other value will lead to major errors across the software.

**Comment by Sh1nto** [comment]
I ran into the same problem with Galera Cluster. If it's mandatory, the suggested solution from NiceGuyIT should be added. I disabled the the auto increment control

```
wsrep_auto_increment_control= off
```

And set Autoincrement 1. Then i did the setup and activated it again. This doesn't work for productive cluster,... so please make a patch.

**Comment by LeDistordu** [comment]
Hi, Are there any plans to patch this problem? Impossible for us to deploy wikijs without a DB cluster, while we want to migrate our mediawiki.

**Reply by NGPixel** [reply]
v3 will no longer rely on hardcoded ID 2 for guests.

**Comment by s-thoeni** [comment]
Sadly I am also unable to use wiki.js because we use a Galera Cluster in our company.

**Comment by bastrian** [comment]
Same issue when using Galera Cluster here. A Patch would really be useful.

edit:
And yes, it will be in v3. But thats still far away.

**Reply by stfast** [reply]
Kind a stupid install in HA enivronment for a project with 25.000+ GitHub stars ....

**Comment by hirschnase** [comment]
Same problem here. MariaDB galera and I can't finish the setup :-/ I even added a special port in HAProxy (3308) which only sends traffic to a single DB node in the cluster but still it's not working.

btw. hard coding ID's is _really_ a bad practice!

**Comment by hirschnase** [comment]
I tried set ```SET GLOBAL wsrep_auto_increment_control = OFF;``` and then do the installation - it looks promising in the first time but then I get an error "The string did not match the expected pattern." and the setup process never finishes.

**Comment by hirschnase** [comment]
FYI what finally worked was the following: I started a stand-alone mariadb server via Docker:

```
docker run -d \
  --name mariadb-temp \
  -e MARIADB_ALLOW_EMPTY_ROOT_PASSWORD=true \
  -e MARIADB_DATABASE=wikijs \
  -e MARIADB_USER=wikijs \
  -e MARIADB_PASSWORD=<PASSWORD> \
  -p 3310:3306 \
  mariadb:latest
  ```
  
  on port 3310 and then pointed WikiJS to it. After the setup ran successfully, I created a DB Dump (mysqldump) of the DB and needed to change the the collation, because I got errors from MariaDB galera that `utf8mb4_uca1400_ai_ci` is not supported:
  
  ```sed -i 's/utf8mb4_uca1400_ai_ci/utf8mb4_general_ci/g' backup.sql```
  
  after I did that, I restored the DB dump into the MariaDB Galera Cluster and pointed WikiJS back to it. After that I was able to login and use it.
  
  it's a mess - but it worked...
